### PR TITLE
man: substantially update the docs regarding hooking sd-bus objects u…

### PR DESCRIPTION
**Pull request by poettering**
…p with external event loops

Prompted by https://lists.freedesktop.org/archives/systemd-devel/2018-December/041817.html

This also drops all references to select() from our manpages. It's 2018
after all, people should use poll(), or ppoll() or epoll().

### CODE REVIEW

File: man/sd_bus_get_fd.xml [hunk 1]
```diff
@@ -34,46 +36,124 @@
...
+    <constant>POLLOUT</constant>, … events, or negative on error.</para>
+
+    <para><function>sd_bus_get_timeout()</function> returns the time-out in µs to pass to to
+    <function>poll()</function> or a similar call when waiting for events on the specified bus connection. The returned
+    time-out may be zero, in which case a subsequent I/O polling call should be invoked in non-blocking mode. The
+    returned timeout may be <constant>UINT64_MAX</constant> in which case the I/O polling call may block indefinitely,
+    without any applied time-out. Note that the returned time-out should be considered only a maximum sleeping time. It
+    is permissible (and even expected) that shorter time-outs are used by the calling program, in case other event
+    sources are polled in the same event loop. Note that the returned time-value is relative and specified in
+    micro-seconds. When converting this value in order to pass it as third argument to <function>poll()</function>
```
Comment (keszybz) [inline man/sd_bus_get_fd.xml]: "microseconds"

File: man/sd_bus_get_fd.xml [hunk 2]
```diff
@@ -34,46 +36,124 @@
...
+    (which expects milli-seconds), care should be taken to use a division that rounds up to ensure the I/O polling
```
Comment (keszybz) [inline man/sd_bus_get_fd.xml]: "milliseconds"

File: man/sd_bus_get_fd.xml [hunk 3]
```diff
@@ -34,46 +36,124 @@
...
+    operation doesn't sleep for shorter than necessary, which might result in unintended busy looping (alternatively,
+    use <citerefentry project='man-pages'><refentrytitle>ppoll</refentrytitle><manvolnum>3</manvolnum></citerefentry>
+    instead of plain <function>poll()</function>, which understands time-outs with nano-second granularity).</para>
+
+    <para>These three functions are useful to hook up a bus connection object with an external or manual event loop
+    involving <function>poll()</function> or a similar I/O polling call. Before each invocation of the I/O polling
+    call, all three functions should be invoked: the file descriptor returned by <function>sd_bus_get_fd()</function>
+    should be polled for the events indicated by <function>sd_bus_get_events()</function>, and the I/O call should
+    block for that up to the time-out returned by <function>sd_bus_get_timeout()</function>. After each I/O polling
+    call the bus connection needs to process incoming or outgoing data, for that call
```
Comment (keszybz) [inline man/sd_bus_get_fd.xml]: for that call → by calling ?

File: man/sd_bus_get_fd.xml [hunk 4]
```diff
@@ -34,46 +36,124 @@
...
+      <varlistentry>
+        <term><constant>-EINVAL</constant></term>
+
+        <listitem><para>An invalid bus object was passed.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><constant>-ECHILD</constant></term>
+
+        <listitem><para>The bus connection was allocated in a parent process and it is being attempted to reuse it in a
```
Comment (keszybz) [inline man/sd_bus_get_fd.xml]: "and is being reused in a child process after fork"?

File: man/sd_bus_process.xml [hunk 5]
```diff
@@ -33,57 +33,100 @@
...
-    <para>
-      Once no further progress can be made,
-      <citerefentry><refentrytitle>sd_bus_wait</refentrytitle><manvolnum>3</manvolnum></citerefentry>
-      should be called. Alternatively the user can wait for incoming data on
-      the file descriptor returned by
-      <citerefentry><refentrytitle>sd_bus_get_fd</refentrytitle><manvolnum>3</manvolnum></citerefentry>.
+    <para><function>sd_bus_process()</function> drives the connection between the client and the message bus. That is,
+    it handles connecting, authentication, and message processing. When invoked pending I/O work is executed, and
+    queued incoming messages are dispatched to registered callbacks. Each time it is invoked a single operation is
+    executed. It returns zero when no operations are pending and positive if one operation was executed. When zero is
```
Comment (keszybz) [inline man/sd_bus_process.xml]: are pending → were pending
one operation was executed → a message was processed ?

File: man/sd_bus_wait.xml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    project='man-pages'><refentrytitle>ppoll</refentrytitle><manvolnum>3</manvolnum></citerefentry>, to wait for I/O on
+    the bus connection. If the <parameter>timeout_sec</parameter> parameter is specified, the call will block at most
+    for the specified amount of time in µs. Pass <constant>UINT64_MAX</constant> to permit it to sleep
+    indefinitely.</para>
+
+    <para>After each invocation of <function>sd_bus_wait()</function> the <function>sd_bus_process()</function> call
+    should be invoked in order to process any now pending I/O work.</para>
+
+    <para>Note that <function>sd_bus_wait()</function> is suitable only for simple programs as it does not permit
+    waiting for other I/O events than the ones of the bus connection itself. For more complex programs either connect
```
Comment (keszybz) [inline man/sd_bus_wait.xml]: " than the ones of the bus connection itself" → ""

File: man/sd_bus_wait.xml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    the bus connection object to an external or manual event loop using
```
Comment (keszybz) [inline man/sd_bus_wait.xml]: "external or manual" → "external"

**Comment by poettering** [comment]
Force pushed a new version. Only changes were the ones @keszybz pointed out (though I replicated the same ECHILD fix to the other two man pages). Taking liberty to set green label, given how minor the fixes are and given that documentation is generally pretty riskless.

**Pull request merged**